File: persistence/src/main/java/io/github/manami/persistence/inmemory/InMemoryPersistenceHandler.java
package io.github.manami.persistence.inmemory;

import io.github.manami.dto.entities.Anime;
import io.github.manami.dto.entities.FilterEntry;
import io.github.manami.dto.entities.MinimalEntry;
import io.github.manami.dto.entities.WatchListEntry;
import io.github.manami.persistence.PersistenceHandler;
import io.github.manami.persistence.inmemory.animelist.InMemoryAnimeListHandler;
import io.github.manami.persistence.inmemory.filterlist.InMemoryFilterListHandler;
import io.github.manami.persistence.inmemory.watchlist.InMemoryWatchListHandler;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.List;
import java.util.UUID;

/**
 * @author manami project
 * @since 2.7.0
 */
@Named("inMemoryStrategy")
public class InMemoryPersistenceHandler implements PersistenceHandler {

    private final InMemoryFilterListHandler filterListHandler;
    private final InMemoryAnimeListHandler animeListHandler;
    private final InMemoryWatchListHandler watchListHandler;


    /**
     * @since 2.7.0
     * @param animeListHandler
     * @param filterListHandler
     * @param watchListHandler
     */
    @Inject
    public InMemoryPersistenceHandler(final InMemoryAnimeListHandler animeListHandler, final InMemoryFilterListHandler filterListHandler, final InMemoryWatchListHandler watchListHandler) {
        this.animeListHandler = animeListHandler;
        this.filterListHandler = filterListHandler;
        this.watchListHandler = watchListHandler;
    }


    @Override
    public boolean filterAnime(final MinimalEntry anime) {
        watchListHandler.removeFromWatchList(anime.getInfoLink());
        return filterListHandler.filterAnime(anime);
    }


    @Override
    public List<FilterEntry> fetchFilterList() {
        return filterListHandler.fetchFilterList();
    }


    @Override
    public boolean filterEntryExists(final String url) {
        return filterListHandler.filterEntryExists(url);
    }


    @Override
    public boolean removeFromFilterList(final String url) {
        return filterListHandler.removeFromFilterList(url);
    }


    @Override
    public boolean addAnime(final Anime anime) {
        filterListHandler.removeFromFilterList(anime.getInfoLink());
        watchListHandler.removeFromWatchList(anime.getInfoLink());
        return animeListHandler.addAnime(anime);
    }


    @Override
    public List<WatchListEntry> fetchWatchList() {
        return watchListHandler.fetchWatchList();
    }


    @Override
    public boolean watchListEntryExists(final String url) {
        return watchListHandler.watchListEntryExists(url);
    }


    @Override
    public boolean watchAnime(final MinimalEntry anime) {
        filterListHandler.removeFromFilterList(anime.getInfoLink());
        return watchListHandler.watchAnime(anime);
    }


    @Override
    public boolean removeFromWatchList(final String url) {
        return watchListHandler.removeFromWatchList(url);
    }


    @Override
    public List<Anime> fetchAnimeList() {
        return animeListHandler.fetchAnimeList();
    }


    @Override
    public boolean animeEntryExists(final String url) {
        return animeListHandler.animeEntryExists(url);
    }


    @Override
    public boolean removeAnime(final UUID id) {
        return animeListHandler.removeAnime(id);
    }


    @Override
    public void clearAll() {
        animeListHandler.clear();
        watchListHandler.clear();
        filterListHandler.clear();
    }


    @Override
    public void addAnimeList(final List<Anime> list) {
        list.forEach(animeListHandler::addAnime);
    }


    @Override
    public void addFilterList(final List<? extends MinimalEntry> list) {
        list.forEach(filterListHandler::filterAnime);
    }


    @Override
    public void addWatchList(final List<? extends MinimalEntry> list) {
        list.forEach(watchListHandler::watchAnime);
    }
}

File: persistence/src/main/kotlin/io/github/manami/persistence/inmemory/InMemoryPersistenceHandler.kt
package io.github.manami.persistence.inmemory

import io.github.manami.dto.entities.*
import io.github.manami.persistence.*
import io.github.manami.persistence.inmemory.animelist.InMemoryAnimeListHandler
import io.github.manami.persistence.inmemory.filterlist.InMemoryFilterListHandler
import io.github.manami.persistence.inmemory.watchlist.InMemoryWatchListHandler


internal class InMemoryPersistenceHandler(
        private val animeListHandler: InMemoryAnimeListHandler,
        private val filterListHandler: InMemoryFilterListHandler,
        private val watchListHandler: InMemoryWatchListHandler
) : ApplicationPersistence, InternalPersistenceHandler {


    override fun filterAnime(anime: MinimalEntry): Boolean {
        return filterListHandler.filterAnime(anime)
    }


    override fun fetchFilterList(): MutableList<FilterListEntry> {
        return filterListHandler.fetchFilterList()
    }


    override fun filterListEntryExists(infoLink: InfoLink): Boolean {
        return filterListHandler.filterListEntryExists(infoLink)
    }


    override fun removeFromFilterList(anime: MinimalEntry): Boolean {
        return filterListHandler.removeFromFilterList(anime)
    }


    override fun addAnime(anime: Anime): Boolean {
        return animeListHandler.addAnime(anime)
    }


    override fun fetchAnimeList(): MutableList<Anime> {
        return animeListHandler.fetchAnimeList()
    }


    override fun animeEntryExists(infoLink: InfoLink): Boolean {
        return animeListHandler.animeEntryExists(infoLink)
    }


    override fun removeAnime(anime: Anime): Boolean {
        return animeListHandler.removeAnime(anime)
    }


    override fun watchAnime(anime: MinimalEntry): Boolean {
        return watchListHandler.watchAnime(anime)
    }


    override fun fetchWatchList(): MutableList<WatchListEntry> {
        return watchListHandler.fetchWatchList()
    }


    override fun watchListEntryExists(infoLink: InfoLink): Boolean {
        return watchListHandler.watchListEntryExists(infoLink)
    }


    override fun removeFromWatchList(anime: MinimalEntry): Boolean {
        return watchListHandler.removeFromWatchList(anime)
    }


    override fun clearAll() {
        animeListHandler.clear()
        watchListHandler.clear()
        filterListHandler.clear()
    }


    override fun addAnimeList(list: MutableList<Anime>) {
        list.forEach { anime -> animeListHandler.addAnime(anime) }
    }


    override fun addFilterList(list: MutableList<FilterListEntry>) {
        list.forEach { anime -> filterListHandler.filterAnime(anime) }
    }


    override fun addWatchList(list: MutableList<WatchListEntry>) {
        list.forEach { anime -> watchListHandler.watchAnime(anime) }
    }


    override fun updateOrCreate(entry: MinimalEntry) {
        when (entry) {
            is Anime -> animeListHandler.updateOrCreate(entry)
            is FilterListEntry -> filterListHandler.updateOrCreate(entry)
            is WatchListEntry -> watchListHandler.updateOrCreate(entry)
        }
    }
}
